### PR TITLE
Support security groups on AWS ELBs. #306

### PR DESCRIPTION
**Pull request by mgulimonov**
Support security groups on AWS ELBs. #306
Please note, that new a dependency added: aws sdk. I think in future it will be much more convinient to use complete Java-friendly API. =)

### CODE REVIEW

**Comment by ckelner** [comment]
https://github.com/TheWeatherCompany/grid-api/issues/306

**Comment by clstokes** [comment]
Hey @mgulimonov, unfortunately the maintainer of the upstream project (https://github.com/greese/dasein-cloud-aws) doesn't want to use the AWS SDK. While I agree it would definitely be easier to use their SDK than the manual XML parsing, it's best that we stay consistent with the upstream project so that we can keep contributing to them and pulling in their changes too. So, please remove the AWS SDK dependency and change this implementation to do the manual XML way? Thanks.

**Comment by mgulimonov** [comment]
Completed. AWS SDK was eliminated.

File: src/main/java/org/dasein/cloud/aws/DelegateRemoteInvocationHandler.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package org.dasein.cloud.aws;
```
Comment (ckelner) [inline src/main/java/org/dasein/cloud/aws/DelegateRemoteInvocationHandler.java]: Hey @mgulimonov - I appreciate your desire to improve this code base, unfortunately changes such as this break patterns established by the maintainer (https://github.com/greese/dasein-cloud-aws/).  Would you mind reverting this?  We'd also like to better understand what you wanted to achieve with these changes, would you mind explaining further?

File: src/main/java/org/dasein/cloud/aws/LogRequest.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package org.dasein.cloud.aws;
```
Comment (ckelner) [inline src/main/java/org/dasein/cloud/aws/LogRequest.java]: See other comment here: https://github.com/TheWeatherCompany/dasein-cloud-aws/pull/81/files#r12327964

File: src/main/java/org/dasein/cloud/aws/network/ElasticLoadBalancer.java [hunk 3]
```diff
@@ -851,6 +844,32 @@ public LoadBalancerHealthCheck createLoadBalancerHealthCheck(@Nonnull LBHealthCh
         }
     }
 
+    @LogRequest
+    @Override
+    public void attachFirewalls(@Nonnull String loadBalancerId, @Nonnull String... firewallIds) throws CloudException, InternalException {
```
Comment (ckelner) [inline src/main/java/org/dasein/cloud/aws/network/ElasticLoadBalancer.java]: Let's rename to setFirewalls

**Comment by clstokes** [comment]
@mgulimonov can you address Chris's comments above so we can finalize these changes?

**Comment by mgulimonov** [comment]
Hey @ckelner 

The main reason of this class was to standardize remote calls, and remove boilerplate code such as APITrace logging and exception handling.

P.S. For this task I'll remove this concept. But I think we should do something to eliminate catastrophic code duplication, for example for new methods we're introducing. In this commit:
https://github.com/TheWeatherCompany/dasein-cloud-aws/commit/7b793fcb94c3718951afdc2a36ea5c7e2e548290
I want to suggest another concept (see EC2Gateway), may be it'll be more convinient.

**Comment by ckelner** [comment]
Hey @mgulimonov understood and I think that both @clstokes and I agree with that sentiment.  I think changes like that probably need to be discussed with the Dasein team.  I'm not sure how we should approach that discussion (e-mail, github issues, irc, hipchat, etc) - I'll defer to @clstokes for guidance if we want to pursue.

Do you mind merging in changes from develop again so we can merge this issue via GitHub?

File: src/main/java/org/dasein/cloud/aws/network/ElasticLoadBalancer.java [hunk 4]
```diff
@@ -1049,6 +1052,31 @@ public LoadBalancerHealthCheck createLoadBalancerHealthCheck(@Nonnull LBHealthCh
         }
     }
 
+    @Override
+    public void setFirewalls(@Nonnull String loadBalancerId, @Nonnull String... firewallIds) throws CloudException, InternalException {
+        ProviderContext ctx = provider.getContext();
```
Comment (clstokes) [inline src/main/java/org/dasein/cloud/aws/network/ElasticLoadBalancer.java]: Add `APITrace.begin()` and `.end()` calls in this method to be consistent with other API calls.

**Pull request merged**